### PR TITLE
assignment002

### PR DESCRIPTION
**Pull request by ayemoekyaw99**

### CODE REVIEW

**Review by thinpyantsoe — APPROVED** [review]
1) php aritsan migrate error
majors table after studens table create
General error: 1824 Failed to open the referenced table 'majors' (Connection: mysql, SQL: alter table `students` add constraint `students_major_id_foreign` 

2)validation error message not show (all blade check)

3) Class "App\Http\Requests\MajorRequest" does not exist (major create and update)

4) major edit.blade where back button click this error show
  App\Http\Controllers\MajorController::edit(): Argument #1 ($id) must be of type int, 
  string given

5) student create
  Class "App\Http\Requests\StudentRequest" does not exist
**why use Request file not exists**

File: assignment_02/app/Contracts/Dao/MajorDaoInterface.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+namespace App\Contracts\Dao;
+
+interface MajorDaoInterface
+{
+    public function getAllMajors();
+    public function create(array $data);
+    public function destroy(int $id);
+    public function find(int $id);
+    public function update(array $data, int $id);
+}
```
Comment (thinpyantsoe) [inline assignment_02/app/Contracts/Dao/MajorDaoInterface.php:12]: Add eof.

File: assignment_02/app/Contracts/Dao/StudentDaoInterface.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+interface StudentDaoInterface
+{
+    public function getAllStudents();
+    public function getAllMajors();
+    public function create(array $data);
+    public function destroy(int $id);
+    public function find(int $id);
+    public function update(array $data, int $id);
+}
```
Comment (thinpyantsoe) [inline assignment_02/app/Contracts/Dao/StudentDaoInterface.php:13]: Add eof.

File: assignment_02/app/Contracts/Services/MajorServiceInterface.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+namespace App\Contracts\Services;
+
+interface MajorServiceInterface
+{
+    public function getAllMajors();
+
+    public function create(array $data);
+
+    public function destroy(int $id);
+    public function getMajorById(int $id);
```
Comment (thinpyantsoe) [inline assignment_02/app/Contracts/Services/MajorServiceInterface.php:12]: If you break with empty line, then break all functions with empty line.
If your don't use new empty line, then remove new lines.
Make all the same format.

File: assignment_02/app/Contracts/Services/StudentServiceInterface.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace App\Contracts\Services;
+
+interface StudentServiceInterface
+{
+    public function getAllStudents();
+    public function getAllMajors();
+    public function create(array $data);
+
```
Comment (thinpyantsoe) [inline assignment_02/app/Contracts/Services/StudentServiceInterface.php:10]: If you break with empty line, then break all functions with empty line.
If your don't use new empty line, then remove new lines.
Make all the same format.

File: assignment_02/app/Dao/MajorDao.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  * Major update
+  
+  * @param mixed $data
+  * @return void
+  */
+    public function update(array $data, int $id): void
+    {
+        $major = Major::find($id);
+        $major->update($data);
+
```
Comment (thinpyantsoe) [inline assignment_02/app/Dao/MajorDao.php:62]: Remove empty line

File: assignment_02/app/Dao/StudentDao.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+   * Student update
+   
+   * @param mixed $data
+   * @return void
+   */
+    public function update(array $data, int $id): void
+    {
+        $student = Student::find($id);
+        $student->update($data);
+
```
Comment (thinpyantsoe) [inline assignment_02/app/Dao/StudentDao.php:72]: Try to use update statement.
Print query log and compare the difference between the two.
For query log
\DB::enableQueryLog();
//Query statement.
\DB::getQueryLog();

File: assignment_02/app/Http/Controllers/ExcelController.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+use App\Models\Student;
+
+class ExcelController extends Controller
+{
+
+
+    /**
+     * @return \Illuminate\Support\Collection
+     */
+    // public function import()
```
Comment (thinpyantsoe) [inline assignment_02/app/Http/Controllers/ExcelController.php:19]: Remove this if no use

File: assignment_02/app/Http/Controllers/StudentController.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\DAO\StudentDAO;
```
Comment (thinpyantsoe) [inline assignment_02/app/Http/Controllers/StudentController.php:5]: DAO folder exist?

**Review by thinpyantsoe — APPROVED** [review]

File: assignment_02/app/Dao/MajorDao.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+     * sample function 
+     *
+     * @return string
+     */
+    public function find($id)
+    {
+        return Major::find($id);
+    }
+
+
```
Comment (ingyinmaycgm) [inline assignment_02/app/Dao/MajorDao.php:51]: Just one line break between functions.

File: assignment_02/app/Http/Controllers/MajorController.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     *show update
+     *
+     *return @string
+     */
+    public function update(MajorRequest $request, int $id)
+    {
+        $this->majorService->update($request->all(), $id);
+        return back()->with(['success' => 'Successful']);
+
```
Comment (ingyinmaycgm) [inline assignment_02/app/Http/Controllers/MajorController.php:86]: Remove empty line.

File: assignment_02/app/Http/Controllers/StudentController.php [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+use Illuminate\Http\Request;
+use App\Http\Requests\StudentRequest;
+use App\Contracts\Services\StudentServiceInterface;
+use ComposerAutoloaderInit9c491b8531eec05ba41a11d9276a5749;
+
+class StudentController extends Controller
+{
+
```
Comment (ingyinmaycgm) [inline assignment_02/app/Http/Controllers/StudentController.php:13]: Remove empty line.

File: assignment_02/app/Http/Controllers/StudentController.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+     *return  @string
+     */
+    public function create()
+    {
+        $majors = $this->studentService->getAllMajors();
+        return view('student.create', compact('majors'));
+    }
+
+    /**
+     *show store
```
Comment (ingyinmaycgm) [inline assignment_02/app/Http/Controllers/StudentController.php:52]: Write document meaningful according to the code you write.

File: assignment_02/app/Models/Major.php [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+use App\Models\Student;
+
+class Major extends Model
+{
+    use HasFactory;
+   protected $fillable = [
+        'name'
+    ];
+
+public function students()
```
Comment (ingyinmaycgm) [inline assignment_02/app/Models/Major.php:16]: Add function document.

File: assignment_02/app/Services/MajorService.php [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * sample function
+     *
+     * @return string
+     */
+    public function getMajorById(int $id)
+    {
+        return $this->majorDao->find($id);
+
```
Comment (ingyinmaycgm) [inline assignment_02/app/Services/MajorService.php:61]: Remove empty line.

File: assignment_02/app/Services/MajorService.php [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * sample function
+     *
+     * @return string
+     */
+    public function destroy(int $id)
+    {
+        return $this->majorDao->destroy($id);
```
Comment (ingyinmaycgm) [inline assignment_02/app/Services/MajorService.php:50]: Need to return?

File: assignment_02/app/Services/MajorService.php [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * sample function
+     *
+     * @return string
+     */
+    public function create(array $data)
+    {
+        return $this->majorDao->create($data);
```
Comment (ingyinmaycgm) [inline assignment_02/app/Services/MajorService.php:40]: Need to return.

File: assignment_02/app/Services/StudentService.php [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+
+    /**
+     * sample function
+     *
+     * @return string
+     */
+    public function create(array $data)
+    {
+        return $this->studentDao->create($data);
```
Comment (ingyinmaycgm) [inline assignment_02/app/Services/StudentService.php:51]: Need to return?

File: assignment_02/app/Services/StudentService.php [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * sample function
+     *
+     * @return string
+     */
+    public function destroy(int $id)
+    {
+        return $this->studentDao->destroy($id);
```
Comment (ingyinmaycgm) [inline assignment_02/app/Services/StudentService.php:61]: Need to return?

File: assignment_02/resources/views/student/edit.blade.php [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+                    @enderror
+                </div>
+
+                <div class="form-group">
+                    <label for="">Address</label>
+                    <textarea class="form-control @error('address') is-invalid @enderror" id="" name="address">{{$student->address}}</textarea>
+                    @error('address')
+                    <div class="invalid-feedback">{{$message}}</div>
+                    @enderror
+                </div>
```
Comment (ingyinmaycgm) [inline assignment_02/resources/views/student/edit.blade.php:64]: These codes seem to be the same as in student/create.blade.
Try to use sub view. 
https://laravel.com/docs/10.x/blade#including-subviews
- Create sub view including common input elements and share that view file from create and eidt blades.

File: assignment_02/resources/views/student/list.blade.php [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+                        <td>{{ $student->id }}</td>
+                        <td>{{ $student->name }}</td>
+
+                        @foreach ($majors as $major)
+                        @if ($student->major_id == $major->id)
+                        <td>
+                            {{ $major->name }}
+                        </td>
+                        @endif
```
Comment (ingyinmaycgm) [inline assignment_02/resources/views/student/list.blade.php:59]: Fix indentation. 
Don't fetch $majors separately. 
Use table join and select major name in $students. 
Or use relationship and fetch via $student->major->name.

**Pull request merged**